### PR TITLE
feat: simplify state.history with to_memory call in micro-agent. Or the call to LLM may exceed the token limit.

### PR DESCRIPTION
**Pull request by assertion**
Currently, we just put the history directly into the prompt, code in CoderAgent:prompt.md:
```
{{ instructions.history_truncated }}
{{ history_to_json(state.history[-10:]) }}
```
but some observation may contains many contents, for example, `BrowserOutputObservation` has `screenshot` which may contains lots of base64 characters, and will make the call to LLM exceeds the token limit.

This pr will call `to_memory` of each `Action` and `Observation` and make the history simpler.

### CODE REVIEW

File: agenthub/micro/agent.py [hunk 1]
```diff
@@ -44,6 +58,7 @@ def step(self, state: State) -> Action:
             state=state,
             instructions=instructions,
             to_json=to_json,
+            history_to_json=history_to_json,
```
Comment (enyst) [inline agenthub/micro/agent.py:61]: I think you're right we should save to_memory(). I just wonder, why wouldn't we just send here the json based on to_memory()?

File: agenthub/micro/agent.py [hunk 2]
```diff
@@ -44,6 +58,7 @@ def step(self, state: State) -> Action:
             state=state,
             instructions=instructions,
             to_json=to_json,
```
Comment (enyst) [inline agenthub/micro/agent.py:60]:
```suggestion
```

File: agenthub/micro/agent.py (hunk 1 above)
Comment (assertion) [reply agenthub/micro/agent.py:61]: You mean just translate the history by calling `to_memory()` outside, and pass the result to `self.prompt_template.render`, so we don't need to call `history_to_json` method when rendering?   

If you mean this, I think this suggestion is ok  to me, what's your opinion @rbren , it seems you're the initial author.

**Comment by enyst** [comment]
Thanks for this, a welcome change.

I have a little question because I seem to remember there's code that attempts to send only objects serialized with `to_memory` to the LLM, so it's curious why there's a remaining `to_json` that doesn't (?) do it. Why wouldn't we do it for everything that gets sent to the LLM from micro? But it's not necessarily important right now.

File: agenthub/micro/agent.py (hunk 1 above)
Comment (enyst) [reply agenthub/micro/agent.py:61]: Perhaps even more, I think there's code in other agents that avoid to save "heavy stuff like screenshots for the user" to `history` in the first place.

Comment (assertion) [reply agenthub/micro/agent.py:61]: Yes, agree. This pr will just fix micro-agent `history` with heavy stuff like screenshots. I suggest to merge this, and later maybe we can refactor the Action/Observation history send to llm strategy, it should have unified processing method.

**Review by enyst — APPROVED** [review]

**Pull request merged**